panels/user-accounts/cc-avatar-chooser.c

```diff
@@ -410,35 +410,61 @@ cheese_camera_device_monitor_new_cb (GObject *source,
 }
 #endif /* HAVE_CHEESE */
 
-static void
-setup_photo_popup (CcAvatarChooser *self)
+static GSList *
+get_settings_facesdirs (void)
 {
-        GFile *file, *dir;
-        GFileInfo *info;
-        GFileEnumerator *enumerator;
-        GFileType type;
-        const gchar *target;
-        const gchar * const * dirs;
-        guint i;
-        gboolean added_faces;
+        GSList *facesdirs = NULL;
+        char **settings_dirs;
+        int i;
 
-        self->faces = g_list_store_new (G_TYPE_FILE);
-        gtk_flow_box_bind_model (GTK_FLOW_BOX (self->flowbox),
-                                 G_LIST_MODEL (self->faces),
-                                 create_face_widget,
-                                 self,
-                                 NULL);
+        g_autoptr(GSettings) settings = NULL;
 
-        g_signal_connect (self->flowbox, "child-activated",
-                          G_CALLBACK (face_widget_activated), self);
+        settings = g_settings_new_with_path ("org.gnome.controlcenter.user-accounts",
+                                             "/org/gnome/controlcenter/user-accounts/");
 
-        dirs = g_get_system_data_dirs ();
-        for (i = 0; dirs[i] != NULL; i++) {
-                char *path;
+        settings_dirs = g_settings_get_strv (settings, "facesdirs");
+        if (settings_dirs != NULL) {
+                for (i = 0; settings_dirs[i] != NULL; i++) {
+                        char *path = settings_dirs[i];
+                        if (path != NULL && g_strcmp0 (path, "") != 0)
+                                facesdirs = g_slist_prepend (facesdirs, g_strdup (path));
+                }
+                g_strfreev (settings_dirs);
+        }
 
-                path = g_build_filename (dirs[i], "pixmaps", "faces", NULL);
+        return g_slist_reverse (facesdirs);
+}
+
+static GSList *
+get_system_facesdirs (void)
+{
+        GSList *facesdirs = NULL;
+        const char * const * data_dirs;
+        int i;
+
+        data_dirs = g_get_system_data_dirs ();
+        for (i = 0; data_dirs[i] != NULL; i++) {
+                char *path = g_build_filename (data_dirs[i], "pixmaps", "faces", NULL);
+                facesdirs = g_slist_prepend (facesdirs, path);
+        }
+
+        return g_slist_reverse (facesdirs);
+}
+
+static gboolean
+add_faces_from_dirs (GListStore *faces, GSList *facesdirs, gboolean add_all)
+{
+        GSList *facesdir_it;
+        gboolean added_faces = FALSE;
+        const gchar *target;
+        GFileType type;
+        GFileInfo *info;
+        GFile *file, *dir;
+        GFileEnumerator *enumerator;
```

Comment:
Optional nitpick: We use C99 declarations-inside-blocks and `g_autoptr` above, so we should either use it consistently in new code, or not use it if the surrounding code doesn't use it, to facilitate upstreaming this; I'm not a gnome-control-center maintainer so I'm not sure what's the right answer.